### PR TITLE
Examples

### PR DESCRIPTION
**Pull request by rakeshv24**
## Changes Made

Added auv_control_demos package.

### CODE REVIEW

File: auv_control_demos/README.md [hunk 1]
```diff
@@ -0,0 +1,321 @@
+# AUV Control Demos
+
+This package contains examples that demonstrate how to use the controllers in
+`auv_controllers`. There are two examples. The first shows how to launch just a
```
Comment (evan-palmer) [inline auv_control_demos/README.md:4]: delete "just"

File: auv_control_demos/README.md [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+single controller on its own. The second example shows how to chain the
```
Comment (evan-palmer) [inline auv_control_demos/README.md:5]: Maybe revise to something like: "The second example demonstrates how to chain the controllers together to construct a cascading control architecture."

File: auv_control_demos/README.md [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+controllers together to a mock hardware interface for an AUV.
+
+## Example 1: Individual Controller
+
+This example shows how to launch an individual auv controller connect them to a
```
Comment (evan-palmer) [inline auv_control_demos/README.md:10]: Use consistent notation: auv -> AUV

Comment (evan-palmer) [inline auv_control_demos/README.md:10]: Grammar

File: auv_control_demos/README.md [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+mock hardware interface. We launch the integral sliding mode controller which
+takes velocity reference and state interfaces and outputs a wrench command
+interface.
```
Comment (evan-palmer) [inline auv_control_demos/README.md:13]: I think you could probably shorten this to just something like: "This example uses the integral sliding mode controller (insert link) to demonstrate how to launch a single controller."

File: auv_control_demos/README.md [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+
+The launch file [individual.launch.py](https://github.com/Robotic-Decision-Making-Lab/auv_controllers/blob/main/auv_control_demos/launch/individual.launch.py)
```
Comment (evan-palmer) [inline auv_control_demos/README.md:15]: Don't really need this sentence

File: auv_control_demos/README.md [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+contains the code to launch the integral sliding mode controller to a mock
+hardware interface, and that is the launch file we use for this first example.
+
+### Tutorial Steps
+
+1. To test the chainable controller framework, run the following command:
```
Comment (evan-palmer) [inline auv_control_demos/README.md:21]: Are you testing the chainable control framework? I thought you were testing an individual controller here?

File: auv_control_demos/README.md [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+
+This package contains examples that demonstrate how to use the controllers in
+`auv_controllers`. There are two examples. The first shows how to launch just a
+single controller on its own. The second example shows how to chain the
+controllers together to a mock hardware interface for an AUV.
+
+## Example 1: Individual Controller
+
+This example shows how to launch an individual auv controller connect them to a
+mock hardware interface. We launch the integral sliding mode controller which
```
Comment (evan-palmer) [inline auv_control_demos/README.md:11]: I don't think you need to mention that this uses a mock hardware interface

File: auv_control_demos/README.md [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+
+### Tutorial Steps
+
+1. To test the chainable controller framework, run the following command:
+
+    ```bash
+    ros2 launch auv_control_demos individual.launch.py
+    ```
+
+    The launch file loads and starts the mock hardware and the controller.
```
Comment (evan-palmer) [inline auv_control_demos/README.md:27]: Reword to "The launch file loads and starts the hardware interface and controller."

File: auv_control_demos/README.md [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+
+2. Check if controllers are running by:
```
Comment (evan-palmer) [inline auv_control_demos/README.md:29]: Grammar

File: auv_control_demos/README.md [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+    ros2 control list_controllers
+    ```
+
+    The output should resemble:
+
+    ```bash
+    integral_sliding_mode_controller[velocity_controllers/IntegralSlidingModeController] active
+    ```
+
+3. Verify if the hardware interfaces are loaded properly by opening another
```
Comment (evan-palmer) [inline auv_control_demos/README.md:41]: "executing the following command"

File: auv_control_demos/README.md (hunk 4 above)
Comment (evan-palmer) [reply auv_control_demos/README.md:13]: Also why chose the ISMC controller for the demo? Could you launch another controller that's maybe a bit easier to use?

File: auv_control_demos/README.md [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+4. Open another terminal and execute the following command:
+
+    ```bash
+    ros2 topic pub /integral_sliding_mode_controller/system_state geometry_msgs/msg/Twist
+    ```
+
+5. In another terminal, execute the following:
+
+    ```bash
+    ros2 run teleop_twist_keyboard teleop_twist_keyboard --ros-args --remap cmd_vel:=/integral_sliding_mode_controller/reference
```
Comment (evan-palmer) [inline auv_control_demos/README.md:82]: I'm concerned about introducing another dependency here just for a demo. If there is an easier way to demonstrate the control output, I think that would be preferable.

File: auv_control_demos/README.md [hunk 12]
```diff
@@ -0,0 +1,321 @@
...
+      reference: 0.0
+      feedback: 0.0
+      feedback_dot: 0.0
+      error: 0.0
+      error_dot: 0.0
+      time_step: 0.004412119
+      output: 0.0
+      ```
+
+  This clearly demonstrates that the integral sliding mode controller is
```
Comment (evan-palmer) [inline auv_control_demos/README.md:152]: Try to refrain from saying "clearly" to users. It can come across as demeaning

File: auv_control_demos/README.md [hunk 13]
```diff
@@ -0,0 +1,321 @@
...
+- Controllers:
+  - [Integral Sliding Mode Controller](https://github.com/Robotic-Decision-Making-Lab/auv_controllers/tree/documentation/velocity_controllers)
+
+- Controller Config:
+  - [individual_controller.yaml](https://github.com/Robotic-Decision-Making-Lab/auv_controllers/blob/main/auv_control_demos/config/individual_controller.yaml)
+
+- URDF File:
+  - [individual.urdf.xacro](https://github.com/Robotic-Decision-Making-Lab/auv_controllers/blob/main/auv_control_demos/urdf/individual.urdf.xacro)
+
+## Example 2: Controller Chaining
```
Comment (evan-palmer) [inline auv_control_demos/README.md:171]: This is a lot for a single README. Could you break this up into two different subdirectories instead of throwing it all into one directory?

File: auv_control_demos/README.md [hunk 14]
```diff
@@ -0,0 +1,321 @@
...
+
+This example shows how to launch the auv controllers chained together and
```
Comment (evan-palmer) [inline auv_control_demos/README.md:173]: auv -> AUV. Might be good to run a find-all & replace for this

File: auv_control_demos/README.md [hunk 15]
```diff
@@ -0,0 +1,321 @@
...
+connect them to a mock hardware interface. This framework takes a
+velocity reference and state interfaces as input and converts them to their
```
Comment (evan-palmer) [inline auv_control_demos/README.md:175]: "This example demonstrates how to launch multiple controllers chained together. The control hierarchy uses a velocity reference input with state feedback and applies a PWM command to the hardware interface."

File: auv_control_demos/README.md [hunk 16]
```diff
@@ -0,0 +1,321 @@
...
+appropriate PWM signal for the hardware interface.
+
+The launch file [chaining.launch.py](https://github.com/Robotic-Decision-Making-Lab/auv_controllers/blob/main/auv_control_demos/launch/chaining.launch.py)
```
Comment (evan-palmer) [inline auv_control_demos/README.md:178]: Don't need this sentence. You mention what launch file you're going to use in the first step of the tutorial below

File: auv_control_demos/README.md [hunk 17]
```diff
@@ -0,0 +1,321 @@
...
+
+### Tutorial Steps
+
+1. To test the chainable controller framework, run the following command:
+
+    ```bash
+    ros2 launch auv_control_demos chaining.launch.py
+    ```
+
+    The launch file loads and starts the mock hardware and the controllers.
```
Comment (evan-palmer) [inline auv_control_demos/README.md:190]: Same comment here as above

File: auv_control_demos/README.md [hunk 18]
```diff
@@ -0,0 +1,321 @@
...
+
+2. Check if controllers are running by:
```
Comment (evan-palmer) [inline auv_control_demos/README.md:192]: Grammar

File: auv_control_demos/README.md [hunk 19]
```diff
@@ -0,0 +1,321 @@
...
+    ros2 topic pub /integral_sliding_mode_controller/system_state geometry_msgs/msg/Twist
+    ```
+
+5. In another terminal, execute the following:
+
+    ```bash
+    ros2 run teleop_twist_keyboard teleop_twist_keyboard --ros-args --remap cmd_vel:=/integral_sliding_mode_controller/reference
+    ```
+
+6. If you echo the `/status' topic of any of the controllers, you would be able
```
Comment (evan-palmer) [inline auv_control_demos/README.md:274]: would -> should

File: auv_control_demos/README.md [hunk 20]
```diff
@@ -0,0 +1,321 @@
...
+      reference: -28.30267593385744
+      feedback: 0.0
+      feedback_dot: 0.0
+      error: 0.0
+      error_dot: 0.0
+      time_step: 0.004444319
+      output: 1182.0
+    ```
+
+  This clearly demonstrates that the controller chaining is functional, as the
```
Comment (evan-palmer) [inline auv_control_demos/README.md:301]: same comment as above

File: auv_control_demos/README.md [hunk 21]
```diff
@@ -0,0 +1,321 @@
...
+  reference velocity commands sent to the integral_sliding_mode_controller are
```
Comment (evan-palmer) [inline auv_control_demos/README.md:302]: `integral_sliding_mode_controller`

File: auv_control_demos/README.md [hunk 22]
```diff
@@ -0,0 +1,321 @@
...
+  converted to the appropriate PWM values as part of the chained controllers and
+  are reflected in the output of the thruster_controller.
```
Comment (evan-palmer) [inline auv_control_demos/README.md:304]: `thruster_controller`

Comment (evan-palmer) [inline auv_control_demos/README.md:304]: This is a run-on sentence

File: auv_control_demos/config/chained_controllers.yaml [hunk 23]
```diff
@@ -0,0 +1,136 @@
+controller_manager:
+  ros__parameters:
+    update_rate: 225 # Hz
```
Comment (evan-palmer) [inline auv_control_demos/config/chained_controllers.yaml:3]: Maybe drop down to like 30 Hz

File: auv_control_demos/config/individual_controller.yaml [hunk 24]
```diff
@@ -0,0 +1,24 @@
+controller_manager:
+  ros__parameters:
+    update_rate: 225 # Hz
```
Comment (evan-palmer) [inline auv_control_demos/config/individual_controller.yaml:3]: Same comment as above

File: auv_control_demos/package.xml [hunk 25]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>auv_control_demos</name>
+  <version>0.0.0</version>
```
Comment (evan-palmer) [inline auv_control_demos/package.xml:5]: Increment version to 0.0.1

File: auv_control_demos/package.xml [hunk 26]
```diff
@@ -0,0 +1,32 @@
...
+  <description>Example package that includes demos for using auv_controllers in individual and chained modes.</description>
```
Comment (evan-palmer) [inline auv_control_demos/package.xml:6]: Delete period. This isn't a sentence

File: auv_control_demos/urdf/chaining.urdf.xacro
Comment (evan-palmer) [inline auv_control_demos/urdf/chaining.urdf.xacro:1]: We still need to clean this up. It's best to split out the ros2_control stuff into its own xacro file, create a URDF file that has the links and joints, and a top-level configuration file. In total, you would have 3 files

File: auv_control_demos/urdf/chaining.urdf.xacro [hunk 27]
```diff
@@ -0,0 +1,199 @@
+<?xml version="1.0"?>
+<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="testbot">
```
Comment (evan-palmer) [inline auv_control_demos/urdf/chaining.urdf.xacro:2]: Maybe something like "auv" instead of "testbot"

File: auv_control_demos/urdf/individual.urdf.xacro
Comment (evan-palmer) [inline auv_control_demos/urdf/individual.urdf.xacro:1]: Same comments here as above

**Review by evan-palmer — CHANGES_REQUESTED** [review]
Thanks for adding this! Mostly just need to do some grammar cleanup and fix the xacro files. Make sure that the CI passes too

File: auv_control_demos/README.md (hunk 1 above)
Comment (rakeshv24) [reply auv_control_demos/README.md:4]: done

File: auv_control_demos/README.md (hunk 2 above)
Comment (rakeshv24) [reply auv_control_demos/README.md:5]: done

File: auv_control_demos/README.md (hunk 3 above)
Comment (rakeshv24) [reply auv_control_demos/README.md:10]: done

Comment (rakeshv24) [reply auv_control_demos/README.md:10]: noted.

File: auv_control_demos/README.md (hunk 4 above)
Comment (rakeshv24) [reply auv_control_demos/README.md:13]: > I think you could probably shorten this to just something like: "This example uses the integral sliding mode controller (insert link) to demonstrate how to launch a single controller."

done

File: auv_control_demos/README.md (hunk 5 above)
Comment (rakeshv24) [reply auv_control_demos/README.md:15]: done

File: auv_control_demos/README.md (hunk 6 above)
Comment (rakeshv24) [reply auv_control_demos/README.md:21]: fixed

File: auv_control_demos/README.md (hunk 7 above)
Comment (rakeshv24) [reply auv_control_demos/README.md:11]: okay.

File: auv_control_demos/README.md (hunk 8 above)
Comment (rakeshv24) [reply auv_control_demos/README.md:27]: done

File: auv_control_demos/README.md (hunk 10 above)
Comment (rakeshv24) [reply auv_control_demos/README.md:41]: fixed

File: auv_control_demos/README.md (hunk 14 above)
Comment (rakeshv24) [reply auv_control_demos/README.md:173]: okay

File: auv_control_demos/README.md (hunk 15 above)
Comment (rakeshv24) [reply auv_control_demos/README.md:175]: fixed

File: auv_control_demos/README.md (hunk 16 above)
Comment (rakeshv24) [reply auv_control_demos/README.md:178]: okay

File: auv_control_demos/README.md (hunk 17 above)
Comment (rakeshv24) [reply auv_control_demos/README.md:190]: okay

File: auv_control_demos/README.md (hunk 4 above)
Comment (EverardoG) [reply auv_control_demos/README.md:13]: We'll make that change. I don't think we have strong reasoning for using the ISMC controller over a different controller for the demo.

I think there is one advantage of showing the demo with this controller as opposed to a simpler controller. If someone wants to try out using a simpler controller, they can omit parts from this demo to run that simpler controller. If we instead show the demo with a simpler controller, then it would probably be confusing trying to figure out what to include to make the demo work with the ISMC controller. That reasoning does go both ways, though.

File: auv_control_demos/README.md (hunk 11 above)
Comment (EverardoG) [reply auv_control_demos/README.md:82]: We could do something like this
ros2 topic pub /integral_sliding_mode_controller/reference geometry_msgs/msg/Twist "{linear: {x: 0.5, y: 0.0, z: 0.0}, angular: {x: 0.0, y: 0.0, z: 0.2}}"

I think it's not as convenient as using the teleop twist keyboard to be able to put in different twist commands with the tap of a key, but it does cut out the extra dependency.

File: auv_control_demos/README.md (hunk 13 above)
Comment (rakeshv24) [reply auv_control_demos/README.md:171]: fixed

File: auv_control_demos/urdf/chaining.urdf.xacro
Comment (rakeshv24) [reply auv_control_demos/urdf/chaining.urdf.xacro:1]: done

**Review by evan-palmer — APPROVED** [review]
Looks great! This is a huge improvement
